### PR TITLE
Support Bech32M, Segwit v1, BIP350

### LINKED ISSUE #1331: Support BIP350; Bech32m address checksum for v1 segwit addresses
https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki 

Currently used v0 segwit addresses use Bech32 checksum.
Upcoming v1+ will use a slightly modified version, Bech32m.

"There should not be any compatibility issues on the receiver side; no wallets are creating v1 segregated witness addresses yet, as the output type is not usable on mainnet."

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Support BIP350, New Bech32M checksum for segwit addresses with version >= 1.  Fixes #1331 .

## Testing instructions

Unit tests (Bech32, Segwit)
## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: src/Bech32.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // Copyright © 2017 Pieter Wuille
-// Copyright © 2017-2020 Trust Wallet.
+// Copyright © 2017-2021 Trust Wallet.
```
Comment (hewigovens) [inline src/Bech32.cpp:2]: let's change default year in `erb` files too?

File: src/Bech32.cpp [hunk 2]
```diff
@@ -62,16 +64,35 @@ Data expand_hrp(const std::string& hrp) {
     return ret;
 }
 
+inline uint32_t xorConstant(ChecksumVariant variant) {
+    if (variant == ChecksumVariant::Bech32) {
+        return 0x01;
```
Comment (hewigovens) [inline src/Bech32.cpp:69]: can you define `0x01` as a const?

File: src/Bech32.cpp [hunk 3]
```diff
@@ -62,16 +64,35 @@ Data expand_hrp(const std::string& hrp) {
...
+}
+
 /** Verify a checksum. */
-bool verify_checksum(const std::string& hrp, const Data& values) {
-    return polymod(cat(expand_hrp(hrp), values)) == 1;
+ChecksumVariant verify_checksum(const std::string& hrp, const Data& values) {
+    Data enc = expand_hrp(hrp);
+    append(enc, values);
+    auto poly = polymod(enc);
+    if (poly == 1) {
```
Comment (hewigovens) [inline src/Bech32.cpp:80]: is this same `0x01`?

File: src/Bech32.h [hunk 4]
```diff
@@ -1,30 +1,45 @@
...
 // file LICENSE at the root of the source code distribution tree.
 
+#include "Data.h"
+
 #include <cstdint>
 #include <string>
 #include <vector>
+#include <tuple>
+
+#include "HexCoding.h" // TODO remove
```
Comment (hewigovens) [inline src/Bech32.h:15]: remove?

**Review by hewigovens — COMMENTED** [review]
looks good, how to use this new address in swift or kotlin?

File: src/Bech32.cpp (hunk 2 above)
Comment (catenocrypt) [reply src/Bech32.cpp]: Done

File: src/Bech32.cpp (hunk 3 above)
Comment (catenocrypt) [reply src/Bech32.cpp]: yes, const added

**Review by hewigovens — APPROVED** [review]

**Review by vikmeup — APPROVED** [review]

**Pull request merged**